### ISSUE TITLE
Number of Routes significantly impacts performance

### ISSUE THREAD

**Issue by Taylor123** [react-native-community/react-native-tab-view#755]
### Current behaviour

I currently have 5 routes (tabs), and the initial tab makes an API (using `redux-api-middleware`) call on `componentDidMount`. Without the `TabView` I have a benchmark of 300 ms from `componentDidMount`. With the 5 routes, the timing takes 1.5s and with 9 routes its about 6 seconds.

It seems that the more routes I have, the worse the JS thread gets locked up and thus takes longer for the full `redux-api-middleware` lifecycle to complete. The degraded performance happens regardless of what is on the routes since I'm lazy loading and ensured they all return null in the `renderScenes` function.

### Expected behaviour

Hoping for better performance, similar to that of [react-native-scrollable-tab-view](https://github.com/ptomasroos/react-native-scrollable-tab-view).

### What have you tried

I've tried all of the performance optimization techniques in the readme.

I benchmarked with and without the `TabView` as well as benchmarking different number of routes.

Is anyone else having issues like this? Any suggestions or ideas on how to improve the performance?

### Your Environment

- `react-native`: `0.59.4`
- `react-native-tab-view`: `2.2.0`
- `react-native-reanimated`: `1.0.1`
- `react-native-gesture-handler`: `1.1.0`

**Comment by satya164** [comment]
Please provide repro

**Comment by wizebin** [comment]
I experienced this same problem because by default this library loads all of the tabs' contents, the easiest solution for me was to use a skeleton page when a route is not loaded and render the actual page when the index changes to it, this of course requires [this open pr](https://github.com/react-native-community/react-native-tab-view/pull/718) to work

This looks something like this:

```javascript
renderRoute = ({ route }) => {
  const { index } = this.state;

  if (index !== route.index) {
    return (
      <SkeletonPage />
    );
  }

  const RouteComponent = route.component;
  if (RouteComponent) return <RouteComponent {...this.props} />
}
```

The biggest problem with loading all of the routes is the quantity of active state listeners increases by quite a bit, and some actions that cause major changes on one page get slowed down on other pages as a result.

**Comment by satya164** [comment]
> I experienced this same problem because by default this library loads all of the tabs' contents, the easiest solution for me was to use a skeleton page when a route is not loaded and render the actual page when the index changes to it, this of course requires this open pr to work

There's a [`lazy`](https://github.com/react-native-community/react-native-tab-view#lazy) prop which renders screens as you need.

There's also [`removeClippedSubviews`](https://github.com/react-native-community/react-native-tab-view#removeclippedsubviews) which can help with memory usage on Android

> The biggest problem with loading all of the routes is the quantity of active state listeners increases by quite a bit

Which listeners do you refer to? How many routes do you have? Number of listeners shouldn't be significant. If it's a problem, you could also remove the listeners from inactive pages.

> and some actions that cause major changes on one page get slowed down on other pages as a result.

By slow down, do you mean the animation is slowed down? What gets slowed down? Can you provide a repro?

**Comment by xaviershay** [comment]
This could be #720 (see most recent comments)

**Comment by wizebin** [comment]
@satya164 

Can't use removeClippedSubview because of the iOS bug `Don't enable this on iOS where this is buggy and views don't re-appear.`

> Which listeners do you refer to?
State listeners, redux or mobx or something else, those components will continue to be updated simply because they're "visible"

> By slow down, do you mean the animation is slowed down?
No, the animations are fine, the slowdown is a whole app slowdown as a result of rendering and updating all of the views at all times, this is not a problem is you don't have a state system like redux. An alternative approach would be stop listening when views are offscreen, either way it requires knowing the current route as soon as possible.

It's possible that lazy handles this, if I have time next week I'll run some tests

**Comment by satya164** [comment]
> the slowdown is a whole app slowdown as a result of rendering and updating all of the views at all times

What exactly slows down? How do you notice this showdown?

Also if few redux listeners are slowing down your app, you probably need to work on optimizing that. Otherwise you're going to run into issues one way or the other, for example inactive screens in stack navigator.

**Comment by booboothefool** [comment]
I am also having some perf issues.
I have lazy loading working - I can see it only fetches/renders the tab after I scroll to it.
However the more tabs I load, the slower the rest of my app gets, I think because it just holds onto them and never lets go. Load 1 => ok, Load 3 => a little slower, Load 6 => grinds to a halt.

**Comment by Taylor123** [comment]
@booboothefool yes the `react-native-scrollable-tab-view` has a significantly better performance. I've actually switched back to that library for the time being, but would prefer to use this library since it's actively maintained.

I ran some perf tests a couple weeks ago, but didn't find any conclusive data as to what the issue is. A large portion of native thread resources were spent on the gesture handling conditional nodes, but i still haven't figured out if/why that would slow down the tab views significantly.
